### PR TITLE
if the users does an advanced checkout to a subdirectory, expose that…

### PR DESCRIPTION
**Pull request by aflat**
… directory as an env variable

when checkout out to a subdirectory, expose that subdir as an env variable

The reason for this pull request is because I use multi-scm to checkout multiple git repos. I need to run commands in those directories, usually git commands on each one. I need to know how each GIT_URL maps to a directory, so I can get more info about that git repo.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.md) doc
- [ ] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [ ] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No findbugs warnings were introduced with my changes
- [x] I have interactively tested my changes
- [ ] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

What types of changes does your code introduce? _Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: src/main/java/hudson/plugins/git/GitSCM.java [hunk 1]
```diff
@@ -1370,6 +1372,16 @@ public void buildEnvironment(Run<?, ?> build, java.util.Map<String, String> env)
                    }
                    env.put(GIT_LOCAL_BRANCH, localBranchName);
                 }
+                
+                RelativeTargetDirectory rtd = getExtensions().get(RelativeTargetDirectory.class);
+                if (rtd != null) {
+
```
Comment (samrocketman) [inline src/main/java/hudson/plugins/git/GitSCM.java]: Trailing white space.

File: src/test/java/hudson/plugins/git/GitSCMTest.java [hunk 2]
```diff
@@ -1672,7 +1672,23 @@ public void testCheckoutSansLocalBranchExtension() throws Exception {
        assertEquals("GIT_BRANCH", "origin/master", getEnvVars(project).get(GitSCM.GIT_BRANCH));
        assertEquals("GIT_LOCAL_BRANCH", null, getEnvVars(project).get(GitSCM.GIT_LOCAL_BRANCH));
     }
+
```
Comment (samrocketman) [inline src/test/java/hudson/plugins/git/GitSCMTest.java]: trailing white space

File: src/test/java/hudson/plugins/git/GitSCMTest.java [hunk 3]
```diff
@@ -1672,7 +1672,23 @@ public void testCheckoutSansLocalBranchExtension() throws Exception {
...
+    public void testCheckoutSansRelativeTargetDirectoryExtension() throws Exception {
+       FreeStyleProject project = setupSimpleProject("master");
 
+       final String commitFile1 = "commitFile1";
+       commit(commitFile1, johnDoe, "Commit number 1");
+       FreeStyleBuild build1 = build(project, Result.SUCCESS, commitFile1);
+
+       assertEquals("GIT_CHECKOUT_DIR", null, getEnvVars(project).get(GitSCM.GIT_CHECKOUT_DIR));
+    }
+
```
Comment (samrocketman) [inline src/test/java/hudson/plugins/git/GitSCMTest.java]: trailing white space

File: src/main/java/hudson/plugins/git/GitSCM.java [hunk 4]
```diff
@@ -1370,6 +1372,16 @@ public void buildEnvironment(Run<?, ?> build, java.util.Map<String, String> env)
                    }
                    env.put(GIT_LOCAL_BRANCH, localBranchName);
                 }
+
```
Comment (samrocketman) [inline src/main/java/hudson/plugins/git/GitSCM.java]: trailing white space.

**Comment by aflat** [comment]
Whitespace removed, missing unit test added

**Comment by aflat** [comment]
Everything look ok now?

File: src/test/java/hudson/plugins/git/GitSCMTest.java [hunk 5]
```diff
@@ -1655,7 +1655,6 @@ public void testCheckoutToDefaultLocalBranch_NULL() throws Exception {
        assertEquals("GIT_BRANCH", "origin/master", getEnvVars(project).get(GitSCM.GIT_BRANCH));
        assertEquals("GIT_LOCAL_BRANCH", "master", getEnvVars(project).get(GitSCM.GIT_LOCAL_BRANCH));
     }
-
```
Comment (MarkEWaite) [inline src/test/java/hudson/plugins/git/GitSCMTest.java]: Please put back this blank line.  Whitespace changes in the diffs are a distraction and can make other pull requests fails with an unnecessary conflict.

File: src/main/java/hudson/plugins/git/GitSCM.java [hunk 6]
```diff
@@ -1388,7 +1398,6 @@ public void buildEnvironment(Run<?, ?> build, java.util.Map<String, String> env)
             }
         }
 
-
```
Comment (MarkEWaite) [inline src/main/java/hudson/plugins/git/GitSCM.java]: Please restore this blank line.  I realize it is an annoyance to have inconsistent formatting in so many places, but the large number of pending pull requests means that smaller changes are less likely to cause one of the pending pull requests to have a whitespace related conflict.

File: src/test/java/hudson/plugins/git/GitSCMTest.java [hunk 7]
```diff
@@ -1672,7 +1671,38 @@ public void testCheckoutSansLocalBranchExtension() throws Exception {
        assertEquals("GIT_BRANCH", "origin/master", getEnvVars(project).get(GitSCM.GIT_BRANCH));
        assertEquals("GIT_LOCAL_BRANCH", null, getEnvVars(project).get(GitSCM.GIT_LOCAL_BRANCH));
     }
+     /**
```
Comment (MarkEWaite) [inline src/test/java/hudson/plugins/git/GitSCMTest.java]: I generally prefer a blank line after the end of a method, if you're willing to add them here.

File: src/main/resources/hudson/plugins/git/GitSCM/buildEnv.properties [hunk 8]
```diff
@@ -3,6 +3,7 @@ GIT_PREVIOUS_COMMIT.blurb=The hash of the commit last built on this branch, if a
 GIT_PREVIOUS_SUCCESSFUL_COMMIT.blurb=The hash of the commit last successfully built on this branch, if any.
 GIT_BRANCH.blurb=The remote branch name, if any.
 GIT_LOCAL_BRANCH.blurb=The local branch name being checked out, if applicable.
+GIT_CHECKOUT_DIR.blurb=The directory that the repository will be checked out to.
```
Comment (MarkEWaite) [inline src/main/resources/hudson/plugins/git/GitSCM/buildEnv.properties]: Could you note in this description that the value is only set if the relative target directory extension is used?  I'd rather not have user complaints that they expected GIT_CHECKOUT_DIR to always be set.

Should it also note that it contains the value assigned to the relative target directory, which is most often a relative path from the workspace directory.

Comment (aflat) [reply src/main/resources/hudson/plugins/git/GitSCM/buildEnv.properties]: I added a comment to say the env var will contain the value in Checkout to a sub-directory. The help for this already talks about it being a relative path, so I would think that part is covered...I hope

File: src/test/java/hudson/plugins/git/GitSCMTest.java (hunk 7 above)
Comment (aflat) [reply src/test/java/hudson/plugins/git/GitSCMTest.java]: added to make it more consistent

File: src/test/java/hudson/plugins/git/GitSCMTest.java (hunk 2 above)
Comment (aflat) [reply src/test/java/hudson/plugins/git/GitSCMTest.java]: I added this back per Mark's request

**Comment by aflat** [comment]
Any other changes needed?

**Comment by aflat** [comment]
Any chance this can get reviewed and/or merged please?

**Pull request merged**